### PR TITLE
MAINT: sparse: non-canonical test cleanup and fixes

### PR DESCRIPTION
**Pull request by perimosocordiae**
I noticed that there were a lot of tests for scipy.sparse marked as known failures, and a lot of confusion regarding what is actually supported vs. what is truly a bug (i.e., #3343, #4409, https://github.com/scipy/scipy/issues/4530#issuecomment-74897379).

This PR attempts to clean that up a bit, by:
- implementing `sum_duplicates()` for the BSR format. This means that all classes with a `sum_duplicates` method now actually sum their duplicates, which simplifies control flow and enables many useful operations for BSR matrices.
- removing all the `knownfailure` tests which actually pass, or could be easily fixed. This necessitated adding a calls to `sum_duplicates` in various methods, so we should be careful about the run-time impact of these changes.
- converting tests concerning counts of `nnz` for non-canonical matrices into skips, rather than known failures. Hopefully this will make it clearer that this behavior is not a bug (it's just annoying).

### CODE REVIEW

**Comment by perimosocordiae** [comment]
The PEP8 issue I missed is now fixed and Travis is green.

I'll admit that this is a big change, so I'm happy to split it up into smaller chunks if desired.

**Comment by rgommers** [comment]
Nice. No need to split I'd think - it's not *that* large.

**Comment by argriffing** [comment]
This looks OK to me, but it will cause trouble with code that is sneaky with the sparse matrix encoding details, for example breaking code that uses a csc graph representation that distinguishes between missing edges vs. edges with value zero. I disagree with the idea that scipy sparse matrices should support that kind of 'masking' use-case, but I'm not sure whether there was any consensus.

Here's an inconsistency in the development branch which I guess is fixed by this PR (both csr and csc will behave like csr below).

```python
>>> from scipy.sparse import csr_matrix, csc_matrix
>>> 
>>> a = sparse.csr_matrix(np.identity(4))
>>> for i, j in product(range(4), repeat=2): a[i, j] = (1 if i == j else 0)
... 
>>> a.nonzero()
(array([0, 1, 2, 3], dtype=int32), array([0, 1, 2, 3], dtype=int32))
>>> 
>>> x = sparse.csc_matrix(np.identity(4))
>>> for i, j in product(range(4), repeat=2): x[i, j] = (1 if i == j else 0)
... 
>>> x.nonzero()
(array([0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3], dtype=int32), array([0, 1, 2, 3, 0, 1, 2, 3, 0, 1, 2, 3, 0, 1, 2, 3], dtype=int32))
```

**Comment by argriffing** [comment]
I hope that @jnothman would look at this before it is merged.

**Comment by perimosocordiae** [comment]
I can confirm that CSC now behaves like CSR in your example with this PR, because the CSC `nonzero` method now correctly handles explicit zeros.

I'm inclined to call the old behavior a bug, but you're right that we should try to find users who were relying on it.

**Comment by perimosocordiae** [comment]
Rebased on master and Travis is happy. @pv, @jnothman: want to take a look?

File: scipy/sparse/bsr.py [hunk 1]
```diff
@@ -486,20 +486,41 @@ def eliminate_zeros(self):
...
+                                         self.indices, mask)
         self.prune()
 
     def sum_duplicates(self):
-        raise NotImplementedError
+        """Eliminate duplicate matrix entries by adding them together
+
+        The is an *in place* operation
+        """
+        self.sort_indices()
```
Comment (pv) [inline scipy/sparse/bsr.py]: Add `if self.has_canonical_format:` handling?

**Comment by pv** [comment]
Based on the discussion in gh-5807, the deduplications could be done in-place

File: scipy/sparse/bsr.py (hunk 1 above)
Comment (perimosocordiae) [inline scipy/sparse/bsr.py]: Done.

File: scipy/sparse/csc.py [hunk 2]
```diff
@@ -164,12 +164,17 @@ def nonzero(self):
...
         row, col = self._swap((major_indices, minor_indices))
 
+        # Remove explicit zeros
+        nz_mask = self.data != 0
+        row = row[nz_mask]
+        col = col[nz_mask]
+
         # Sort them to be in C-style order
-        ind = np.lexsort((col, row))
+        ind = np.argsort(row, kind='mergesort')
```
Comment (pv) [inline scipy/sparse/csc.py]: Does this require `self.has_sorted_indices`?

**Comment by pv** [comment]
One remaining question, otherwise LGTM

File: scipy/sparse/csc.py (hunk 2 above)
Comment (perimosocordiae) [inline scipy/sparse/csc.py]: As I understand it, `col` will always be in sorted order here, as it gets generated by `_sparsetools.expandptr`, so we only need to stable-sort by `row` to ensure the correct ordering.

**Comment by perimosocordiae** [comment]
Thanks, @pv!

**Pull request merged**